bundle/Service/ScheduledVisibilityService.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Netgen\Bundle\IbexaScheduledVisibilityBundle\Service;
+
+use DateTime;
+use Ibexa\Contracts\Core\Repository\Values\Content\Content;
+use Ibexa\Contracts\Core\Repository\Values\Content\Field;
+use Ibexa\Core\FieldType\Date\Value as DateValue;
+use Ibexa\Core\FieldType\DateAndTime\Value as DateAndTimeValue;
+use Netgen\Bundle\IbexaScheduledVisibilityBundle\Enums\VisibilityUpdateResult;
+use Netgen\Bundle\IbexaScheduledVisibilityBundle\ScheduledVisibility\Registry;
+use OutOfBoundsException;
+
+final class ScheduledVisibilityService
+{
+    public function __construct(
+        private readonly Registry $registry,
+    ) {}
+
+    /**
+     * @throws OutOfBoundsException
+     */
+    public function updateVisibilityIfNeeded(Content $content, ?string $handlerIdentifier = null): VisibilityUpdateResult
+    {
+        if (!$this->accept($content)) {
+            return VisibilityUpdateResult::NoChange;
+        }
+
+        $handler = $this->registry->get($handlerIdentifier);
+
+        if ($this->shouldBeHidden($content) && !$handler->isHidden($content)) {
+            $handler->hide($content);
+
+            return VisibilityUpdateResult::Hidden;
+        }
+
+        if ($this->shouldBeVisible($content) && $handler->isHidden($content)) {
+            $handler->reveal($content);
+
+            return VisibilityUpdateResult::Revealed;
+        }
+
+        return VisibilityUpdateResult::NoChange;
+    }
+
+    public function accept(Content $content): bool
```

Comment:
This method can now be made private and renamed to `isValid()`. It should only check whether the fields are present, are of the expected field type and the "to" date is greater than "from" date if not empty.

In `updateVisibilityIfNeeded()` we should call the method on top and throw new `InvalidStateException` if the given `Content` is not valid. All consuming code should catch that exception and log an `error` level message.

Reply:
fixed in cde1a7391eb2eb37a59170efd5fad9d64f3684bf